fdbserver/KeyValueStoreMemory.actor.cpp

```diff
@@ -504,56 +519,95 @@ class KeyValueStoreMemory final : public IKeyValueStore, NonCopyable {
-			log->push(StringRef((const uint8_t*)&cipherHeader, BlobCipherEncryptHeader::headerSize));
-			log->push(ciphertext);
+			if (CLIENT_KNOBS->ENABLE_CONFIGURABLE_ENCRYPTION) {
+				BlobCipherEncryptHeaderRef headerRef;
+				StringRef cipherText = cipher.encrypt(plaintext, v1.size() + v2.size(), &headerRef, arena);
+				Standalone<StringRef> headerRefStr = BlobCipherEncryptHeaderRef::toStringRef(headerRef);
+				headerSize = headerRefStr.size();
+				ASSERT(headerSize > 0);
+				log->push(StringRef((const uint8_t*)&headerSize, sizeof(headerSize)));
+				log->push(headerRefStr);
+				log->push(cipherText);
+			} else {
+				BlobCipherEncryptHeader cipherHeader;
+				StringRef ciphertext =
+				    cipher.encrypt(plaintext, v1.size() + v2.size(), &cipherHeader, arena)->toStringRef();
+				headerSize = BlobCipherEncryptHeader::headerSize;
+				ASSERT(headerSize > 0);
+				log->push(StringRef((const uint8_t*)&headerSize, sizeof(headerSize)));
+				log->push(StringRef((const uint8_t*)&cipherHeader, headerSize));
+				log->push(ciphertext);
+			}
 		}
 		return log->push("\x01"_sr); // Changes here should be reflected in OP_DISK_OVERHEAD
 	}
 
 	// In case the op data is not encrypted, simply read the operands and the zero fill flag.
 	// Otherwise, decrypt the op type and data.
 	ACTOR static Future<Standalone<StringRef>> readOpData(KeyValueStoreMemory* self,
-	                                                      OpHeader* h,
+	                                                      OpHeader h,
 	                                                      bool* isZeroFilled,
-	                                                      int* zeroFillSize) {
+	                                                      int* zeroFillSize,
+	                                                      bool encryptedOp) {
 		// Metadata op types to be excluded from encryption.
 		static std::unordered_set<OpType> metaOps = { OpSnapshotEnd, OpSnapshotAbort, OpCommit, OpRollback };
-		if (metaOps.count((OpType)h->op) == 0) {
+		if (metaOps.count((OpType)h.op) == 0) {
 			// It is not supported to open an encrypted store as unencrypted, or vice-versa.
-			ASSERT_EQ(h->op == OpEncrypted, self->enableEncryption);
+			ASSERT_EQ(encryptedOp, self->enableEncryption);
 		}
-		state int remainingBytes = h->len1 + h->len2 + 1;
-		if (h->op == OpEncrypted) {
+		// if encrypted op read the header size
+		state uint16_t headerSize = 0;
+		if (encryptedOp) {
+			state Standalone<StringRef> headerSizeStr = wait(self->log->readNext(sizeof(headerSize)));
+			ASSERT(headerSizeStr.size() <= sizeof(headerSize));
+			// Partial read on the header size
+			memset(&headerSize, 0, sizeof(headerSize));
+			memcpy(&headerSize, headerSizeStr.begin(), headerSizeStr.size());
+			if (headerSizeStr.size() < sizeof(headerSize)) {
+				CODE_PROBE(true, "zero fill partial encryption header size", probe::decoration::rare);
```

Comment:
Is this a valid case?

Reply:
Yeah if we do a partial read, a similar check is done for the `OpHeader` as well before this

Reply:
Am not sure if we should support it, from the encryption and security point of view, it doesn't make much sense?

Reply:
For instance: if `authenticated encryption` is enabled; authToken match would fail for such a read...

Reply:
It's not something we should support it is something we have to support

Reply:
There is a possibility of partial commits to the TxnStateStore so we need to fix a partial commit by appending zeros to the TxnStateStore

Reply:
> It's not something we should support it is something we have to support

Lets take this offline, my understanding of above comment is, we need to support this usecase. However, my point is if EncryptionHeader is partially written, why it can't be taken as a `attack` ? 
Thanks

Reply:
There is subsequent code that if entries are 0 it will not process them, so we would never try to decrypt it, we just need to make sure the commit is not partial (that is filled with 0 for whatever is missing)

Reply:
Lets talk about it, if thats the case, it would be better to understand why do we even need to write this at the first place (both CPU and disk-space can be saved..)

Reply:
@sfc-gh-ahusain The KVStoreMemory is very old code. From my read of the code, the assumption is that it is backed by an append-only on disk log file. And for the write to that file, uncommitted changes can get truncated (e.g. if the last write is un-synced). If it gets a file corrupted at the end, it still want to reuse the file to append more data, and if another recover happen, the file should allow to be read again (i.e. only have corruption at the end). And so the logic here handles the case, fixing the corrupted file by appending \x0 bytes to complete the last truncated record.

All the above have nothing to do with txnStateStore and TLog. The txnStateStore just reuse the whole implementation, and the TLog is plugged-in with a wrapper to simulate an append-only log file.

Reply:
> However, my point is if EncryptionHeader is partially written, why it can't be taken as a attack ?

It would be interesting to know how to construct an attack in this case. I think is it the same with the rest of our encryption-at-rest implementation, where even if an attacker gain control to the disk, confidentiality is guaranteed (and if authentication is enabled, fidelity is guaranteed).

Reply:
@sfc-gh-yiwu, thanks for great explanation. 
Just for clarity: do you know the granularity of partial commits; can a mutation be split between two files? If yes, am not sure how would it work when the mutation is supposed to be encypted?

Reply:
Talked offline with Nim and Yi. Seems like the current implementation doesn't support log truncation, hence, not much that could be done.